src/Agent/NewRelic/Agent/Extensions/NewRelic.Agent.Extensions/Helpers/AwsSdk.cs

```diff
@@ -0,0 +1,128 @@
+{
+    public static class AwsSdkHelpers
+    {
+        private static Regex RegionRegex = new Regex(@"^[a-z]{2}((-gov)|(-iso([a-z]?)))?-[a-z]+-\d{1}$", RegexOptions.Compiled);
+        private static bool LooksLikeARegion(string text) => RegionRegex.IsMatch(text);
+        private static bool LooksLikeAnAccountId(string text) => (text.Length == 12) && text.All(c => c >= '0' && c <= '9');
+        // Only log ARNs we can't parse once
+        private static HashSet<string> BadInvocations = new HashSet<string>();
+
+        // This is the full regex pattern for an ARN:
+        // (arn:(aws[a-zA-Z-]*)?:lambda:)?([a-z]{2}((-gov)|(-iso([a-z]?)))?-[a-z]+-\d{1}:)?(\d{12}:)?(function:)?([a-zA-Z0-9-_\.]+)(:(\$LATEST|[a-zA-Z0-9-_]+))?
+
+        // If it's a full ARN, it has to start with 'arn:'
+        // A partial ARN can contain up to 5 segments separated by ':'
+        // 1. Region
+        // 2. Account ID
+        // 3. 'function' (fixed string)
+        // 4. Function name
+        // 5. Alias or version
+        // Only the function name is required, the rest are all optional. e.g. you could have region and function name and nothing else
+        public static string ConstructArn(IAgent agent, string invocationName, string region, string accountId)
+        {
+            if (invocationName.StartsWith("arn:"))
+            {
+                return invocationName;
+            }
+            var segments = invocationName.Split(':');
+            string functionName = null;
+            string alias = null;
+            string fallback = null;
+
+            foreach (var segment in segments)
+            {
+                if (LooksLikeARegion(segment))
+                {
+                    if (string.IsNullOrEmpty(region))
+                    {
+                        region = segment;
+                    }
+                    else
+                    {
+                        fallback = segment;
+                    }
+                    continue;
+                }
+                else if (LooksLikeAnAccountId(segment))
+                {
+                    if (string.IsNullOrEmpty(accountId))
+                    {
+                        accountId = segment;
+                    }
+                    else
+                    {
+                        fallback = segment;
+                    }
+                    continue;
+                }
+                else if (segment == "function")
+                {
+                    continue;
+                }
+                else if (functionName == null)
+                {
+                    functionName = segment;
+                }
+                else if (alias == null)
+                {
+                    alias = segment;
+                }
+                else
+                {
+                    if (BadInvocations.Add(invocationName))
+                    {
+                        agent?.Logger.Debug($"Unable to parse function name '{invocationName}'");
```

Comment:
1. It seems like it may be possible to run this code concurrently (two async lambda invocations), and we are not using a thread safe collection. This might cause some problems.
2. Since we do not log this at the default log level, we may want to perform a is log level enabled check before storing that invocationName indefinitely.
3. Do we need to put a cap on this collection so that it doesn't grow too large? Should we just log the first N times we encounter this problem?

Reply:
This comment applies to the other places where we use this collection and log a message.

Reply:
Would supportability metrics be useful for debugging these types of problems?

Reply:
1. Ideally we're never going to hit this, but that is technically possible. I can make it thread safe.
2. Sure
3. Since it's a Set, they'd have to be invoking a significant number of unique Lambda names, all of which are un-parse-able. That seems pretty unlikely, but a size check is easy enough.
4. I've replaced all the logging calls with a helper method.
5. I don't think so. This is more for troubleshooting individual cases of "my Lambda calls aren't linking". If we saw a large number of parsing failures, we'd still need to see some concrete examples in order to fix it. AWS provides a regex for what's a valid function identifier, and I've tried to cover all the possibilities in the unit tests, so I'm not expecting many surprises.

Reply:
If we don't think this is likely, and we just need to capture a few examples, maybe we don't need a collection at all and just log the first 10 times we get something we can't parse?